src/Database/LSMTree/Internal/WriteBuffer.hs

```diff
@@ -18,4 +20,139 @@
+
+data Entry v blobref
+    = Insert !v
+    | InsertWithBlob !v !blobref
+    | Mupdate !v
+    | Delete
+  deriving (Show, Functor, Foldable, Traversable)
+
+newtype WriteBuffer k v blobref = WB (Map k (Entry v blobref))
+
+emptyWriteBuffer :: WriteBuffer k v blobref
+emptyWriteBuffer = WB Map.empty
+
+{-------------------------------------------------------------------------------
+  Updates
+-------------------------------------------------------------------------------}
+
+combine :: (v -> v -> v) -> Entry v blobref -> Entry v blobref -> Entry v blobref
+combine _ e@Delete            _                       = e
+combine _ e@Insert {}         _                       = e
+combine _ e@InsertWithBlob {} _                       = e
+combine _   (Mupdate u)       Delete                  = Insert u
+combine f   (Mupdate u)       (Insert v)              = Insert (f u v)
+combine f   (Mupdate u)       (InsertWithBlob v blob) = InsertWithBlob (f u v) blob
+combine f   (Mupdate u)       (Mupdate v)             = Mupdate (f u v)
+
+addEntryMonoidal :: Ord k
+  => (v -> v -> v) -- ^ merge function
+  -> k -> Monoidal.Update v -> WriteBuffer k v blobref -> WriteBuffer k v blobref
+addEntryMonoidal f k e (WB wb) = WB (Map.insertWith (combine f) k (g e) wb) where
+  g :: Monoidal.Update v -> Entry v blobref
+  g (Monoidal.Insert v)  = Insert v
+  g (Monoidal.Mupsert v) = Mupdate v
+  g (Monoidal.Delete)    = Delete
+
+addEntryNormal :: Ord k
+  => k -> Normal.Update v blobref -> WriteBuffer k v blobref -> WriteBuffer k v blobref
+addEntryNormal k e (WB wb) = WB (Map.insert k (g e) wb) where
+  g :: Normal.Update v blobref -> Entry v blobref
+  g (Normal.Insert v Nothing)     = Insert v
+  g (Normal.Insert v (Just bref)) = InsertWithBlob v bref
+  g Normal.Delete                 = Delete
+
+{-------------------------------------------------------------------------------
+  Querying
+-------------------------------------------------------------------------------}
+
+lookupsNormal :: forall k v blobref. Ord k
+  => WriteBuffer k v blobref
+  -> [k]
+  -> [Normal.LookupResult k v blobref]
+lookupsNormal (WB m) = fmap f where
+  f :: k -> Normal.LookupResult k v blobref
+  f k = case Map.lookup k m of
+    Nothing                      -> Normal.NotFound k
+    Just Delete                  -> Normal.NotFound k
+    Just (Insert v)              -> Normal.Found k v
+    Just (InsertWithBlob v bref) -> Normal.FoundWithBlob k v bref
+    Just (Mupdate _v)            -> error "should not happen"
+
+-- | In monoidal case we cannot return 'Monoidal.LookupResult',
+-- as we need to return the partial 'MUpdate' if it's in the writebuffer.
+--
+-- Note: the entry may be 'Delete'.
+--
+-- TODO: create a new data?
+lookupsMonoidal :: forall k v blobref. Ord k
+  => WriteBuffer k v blobref
+  -> [k]
+  -> [(k, Maybe (Entry v blobref))]
+lookupsMonoidal (WB m) = fmap f where
+   f :: k -> (k, Maybe (Entry v blobref))
+   f k = (k, Map.lookup k m)
```

Comment:
Right, we will need to return intermediate results rather than final `LookupResult`, both for monoidal and even for normal.

Ultimately we'll combine the lookup from the write buffer with lookups from the other disk runs and combine them together into a `LookupResult`.

Consider for example in the normal case, if there's no entry for a key, but there is an entry in a later disk run, we would loose information if we return `NotFound` both for no entry and for finding a delete entry.

I think the right thing to do is indeed to return `Entry`, both for normal and monoidal. So it can be a single implementation for both cases. I think what we'll do later is promote the `Entry` type (or equivalent) to be the type used for lookup results from the disk files. The top level lookup impl will combine the results from the write buffer with the results from the disk runs and convert to the appropriate `LookupResult` external API type.

So concretely, I think we just need a single `lookupEntries`
```
lookupEntries :: forall k v blobref. Ord k
  => WriteBuffer k v blobref
  -> [k]
  -> [Maybe (Entry v blobref)]
lookupsMonoidal (WB m) = map (\k -> Map.lookup k m)
```